crates/ide_assists/src/handlers/extract_function.rs

```diff
@@ -643,6 +643,16 @@ fn vars_used_in_body(ctx: &AssistContext, body: &FunctionBody) -> Vec<Local> {
         .collect()
 }
 
+fn body_contains_await(body: &FunctionBody) -> bool {
+    body.descendants()
+        .find(|d| {
+            d.descendants_with_tokens()
+                .find(|d| matches!(d.kind(), SyntaxKind::AWAIT_EXPR | SyntaxKind::AWAIT_KW))
+                .is_some()
+        })
+        .is_some()
+}
```

Comment:
This won't work properly for macros. Assuming a macro that generates an expression using `.await` this will still return false, likewise a macro that gets `await` passed into it without emitting an actual await expression will incorrectly return true.
I don't think we have something like `Semantics::descendants_with_macros` though to look for await expressions 🤔

Reply:
I couldn't see something like that, I have attempted to write one iterating over the descendants and when encountering a  `MacroCall` expanding it with `Semantics::expand` into the tree although I haven't got that to work yet. I'll carry on looking at that as I feel I'm getting more familiar with the code now.

Reply:
@matklad maybe you got an idea here?

Reply:
It does seem like this kind of thing should be done on the HIR, but we need a HIR for expressions for that (independently of #8713).